common/common.h

```diff
@@ -595,30 +596,43 @@ struct common_chat_msg {
     std::string content;
 };
 
-// Get the built-in chat template for the model. Return empty string if not present.
-std::string common_get_builtin_chat_template(const struct llama_model * model);
-
 // Check if the template supplied via "--chat-template" is supported or not. Returns true if it's valid
-bool common_chat_verify_template(const std::string & tmpl);
+bool common_chat_verify_template(const std::string & tmpl, bool use_jinja);
+
+namespace minja {
+    class chat_template;
+}
+
+typedef minja::chat_template llama_chat_template;
+
+struct llama_chat_templates {
+    bool has_explicit_template; // Model had builtin template or template overridde was specified.
+    std::unique_ptr<llama_chat_template> default_template; // always set (defaults to chatml)
```

Comment:
If this is always set, I would prefer having the constructor of `llama_chat_templates` to initialize it. This way, you can get rid of this `std::unique_ptr` for `default_template`

AFAIU `minja::chat_template` is move-able but don't have default constructor, so I think doing explicitly `std::move` would make it safer.

But feel free to ignore if it implies too many changes. This can be done in another PR.

Reply:
Here is a small demo on what I'm talking about: https://github.com/ochafik/llama.cpp/compare/jinja...ggerganov:llama.cpp:xsn/tmp_jinja_safer?expand=1

Reply:
Second option is:

```cpp
struct llama_chat_templates {
    bool has_explicit_template; // Model had builtin template or template overridde was specified.
    llama_chat_template default_template; // always set (defaults to chatml)
    std::unique_ptr<llama_chat_template> tool_use_template;
};

    llama_chat_template default_template(default_template_src, bos_token, eos_token);
    return {
        has_explicit_template,
        std::move(default_template),
        tool_use_template_src.empty()
            ? nullptr
            : std::make_unique<minja::chat_template>(tool_use_template_src, bos_token, eos_token)
    };
```

Reply:
Thanks for the suggestion & test. I actually had something like this initially but switched to unique_ptr to support forward-declaring llama_chat_template (= minja::chat_template) (cf. [earlier review comment](https://github.com/ggerganov/llama.cpp/pull/11016#discussion_r1921004026))

Reply:
(looking to add a non-inlined ctor to enforce the non-nullability of the default template)

Reply:
> (looking to add a non-inlined ctor to enforce the non-nullability of the default template)

actually not so easy now that chat_templates is a field of server_context, leaving as is for now.